src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,283 @@
+                    }
+
+                    if (type.IsValueType && TypeIsAutoLayoutOrContainsAutoLayout(type))
+                    {
+                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ByRefParameters));
+                    }
+                }
+            }
+
+            private bool TypeIsAutoLayoutOrContainsAutoLayout(ITypeSymbol type)
+            {
+                Debug.Assert(type.IsValueType);
+
+                try
+                {
+                    _layoutCacheLock.EnterReadLock();
+                    if (_isAutoLayoutOrContainsAutoLayoutCache.TryGetValue(type, out bool isAutoLayoutOrContainsAutoLayout))
+                    {
+                        return isAutoLayoutOrContainsAutoLayout;
+                    }
+                }
+                finally
+                {
+                    _layoutCacheLock.ExitReadLock();
+                }
+
+                if (_structLayoutAttribute is not null)
+                {
+                    foreach (var attr in type.GetAttributes())
+                    {
+                        if (attr.AttributeClass.OriginalDefinition.Equals(_structLayoutAttribute))
+                        {
+                            var argument = attr.ConstructorArguments[0];
+                            if (argument.Type != null)
+                            {
+                                SpecialType specialType = argument.Type.TypeKind == TypeKind.Enum ?
+                                    ((INamedTypeSymbol)argument.Type).EnumUnderlyingType.SpecialType :
+                                    argument.Type.SpecialType;
+
+                                if (DiagnosticHelpers.TryConvertToUInt64(argument.Value, specialType, out ulong convertedLayoutKindValue) &&
+                                    convertedLayoutKindValue == (ulong)LayoutKind.Auto)
+                                {
+                                    try
+                                    {
+                                        _layoutCacheLock.EnterWriteLock();
+                                        _isAutoLayoutOrContainsAutoLayoutCache.Add(type, true);
+                                        return true;
+                                    }
+                                    finally
+                                    {
+                                        _layoutCacheLock.ExitWriteLock();
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+
+                foreach (var member in type.GetMembers())
+                {
+                    if (member is IFieldSymbol { IsStatic: false, Type.IsValueType: true } valueTypeField)
+                    {
+                        return TypeIsAutoLayoutOrContainsAutoLayout(valueTypeField.Type);
+                    }
+                }
```

Comment:
Is it expected we return on the first non-static value type field without looking into the remaining fields?